File: api/algorithms/data-structures/implementations/python/graph.py
"""
Graph Implementation in Python
Nodes (vertices) connected by edges
Time Complexity: O(V + E) for traversal
"""

class Graph:
    def __init__(self, directed=False):
        self.vertices = {}
        self.directed = directed

    def add_vertex(self, vertex):
        """Add vertex - O(1)"""
        if vertex not in self.vertices:
            self.vertices[vertex] = []
        return self

    def add_edge(self, vertex1, vertex2, weight=1):
        """Add edge - O(1)"""
        if vertex1 not in self.vertices:
            self.add_vertex(vertex1)
        if vertex2 not in self.vertices:
            self.add_vertex(vertex2)

        self.vertices[vertex1].append({'vertex': vertex2, 'weight': weight})

        if not self.directed:
            self.vertices[vertex2].append({'vertex': vertex1, 'weight': weight})

        return self

    def get_neighbors(self, vertex):
        """Get neighbors - O(1)"""
        return self.vertices.get(vertex, [])

    def dfs(self, start_vertex, visited=None, result=None):
        """Depth-First Search (DFS) - O(V + E)"""
        if visited is None:
            visited = set()
        if result is None:
            result = []
        
        visited.add(start_vertex)
        result.append(start_vertex)

        neighbors = self.get_neighbors(start_vertex)
        for neighbor in neighbors:
            if neighbor['vertex'] not in visited:
                self.dfs(neighbor['vertex'], visited, result)

        return result

    def dfs_iterative(self, start_vertex):
        """DFS Iterative - O(V + E)"""
        visited = set()
        result = []
        stack = [start_vertex]

        while stack:
            vertex = stack.pop()

            if vertex not in visited:
                visited.add(vertex)
                result.append(vertex)

                neighbors = self.get_neighbors(vertex)
                for neighbor in reversed(neighbors):
                    if neighbor['vertex'] not in visited:
                        stack.append(neighbor['vertex'])

        return result

    def bfs(self, start_vertex):
        """Breadth-First Search (BFS) - O(V + E)"""
        visited = set()
        result = []
        queue = [start_vertex]
        visited.add(start_vertex)

        while queue:
            vertex = queue.pop(0)
            result.append(vertex)

            neighbors = self.get_neighbors(vertex)
            for neighbor in neighbors:
                if neighbor['vertex'] not in visited:
                    visited.add(neighbor['vertex'])
                    queue.append(neighbor['vertex'])

        return result

    def has_path(self, start_vertex, end_vertex):
        """Check if path exists - O(V + E)"""
        visited = set()
        queue = [start_vertex]
        visited.add(start_vertex)

        while queue:
            vertex = queue.pop(0)

            if vertex == end_vertex:
                return True

            neighbors = self.get_neighbors(vertex)
            for neighbor in neighbors:
                if neighbor['vertex'] not in visited:
                    visited.add(neighbor['vertex'])
                    queue.append(neighbor['vertex'])

        return False

    def get_vertices(self):
        """Get all vertices - O(V)"""
        return list(self.vertices.keys())

    def get_edges(self):
        """Get all edges - O(V + E)"""
        edges = []
        for vertex, neighbors in self.vertices.items():
            for neighbor in neighbors:
                edges.append({
                    'from': vertex,
                    'to': neighbor['vertex'],
                    'weight': neighbor['weight']
                })
        return edges

    def get_vertex_count(self):
        """Get vertex count - O(1)"""
        return len(self.vertices)

    def is_empty(self):
        """Check if empty - O(1)"""
        return len(self.vertices) == 0

    def clear(self):
        """Clear graph - O(1)"""
        self.vertices.clear()


# Usage Example
if __name__ == "__main__":
    graph = Graph(False)  # Undirected graph

    graph.add_vertex('A')
    graph.add_vertex('B')
    graph.add_vertex('C')
    graph.add_vertex('D')
    graph.add_edge('A', 'B')
    graph.add_edge('A', 'C')
    graph.add_edge('B', 'D')
    graph.add_edge('C', 'D')

    print('Vertices:', graph.get_vertices())
    print('DFS:', graph.dfs('A'))
    print('BFS:', graph.bfs('A'))
    print('Has path A to D:', graph.has_path('A', 'D'))
    print('Neighbors of A:', graph.get_neighbors('A'))

File: api/algorithms/data-structures/implementations/java/Graph.java
/**
 * Graph Implementation in Java
 * Nodes (vertices) connected by edges
 * Time Complexity: O(V + E) for traversal
 */
import java.util.*;

public class Graph {
    private Map<String, List<Edge>> vertices;
    private boolean directed;

    private static class Edge {
        String vertex;
        int weight;

        Edge(String vertex, int weight) {
            this.vertex = vertex;
            this.weight = weight;
        }
    }

    public Graph(boolean directed) {
        this.vertices = new HashMap<>();
        this.directed = directed;
    }

    public Graph() {
        this(false);
    }

    public Graph addVertex(String vertex) {
        if (!vertices.containsKey(vertex)) {
            vertices.put(vertex, new ArrayList<>());
        }
        return this;
    }

    public Graph addEdge(String vertex1, String vertex2, int weight) {
        if (!vertices.containsKey(vertex1)) {
            addVertex(vertex1);
        }
        if (!vertices.containsKey(vertex2)) {
            addVertex(vertex2);
        }

        vertices.get(vertex1).add(new Edge(vertex2, weight));

        if (!directed) {
            vertices.get(vertex2).add(new Edge(vertex1, weight));
        }

        return this;
    }

    public Graph addEdge(String vertex1, String vertex2) {
        return addEdge(vertex1, vertex2, 1);
    }

    public List<Edge> getNeighbors(String vertex) {
        return vertices.getOrDefault(vertex, new ArrayList<>());
    }

    public List<String> dfs(String startVertex) {
        List<String> result = new ArrayList<>();
        Set<String> visited = new HashSet<>();
        dfsRecursive(startVertex, visited, result);
        return result;
    }

    private void dfsRecursive(String vertex, Set<String> visited, List<String> result) {
        visited.add(vertex);
        result.add(vertex);

        List<Edge> neighbors = getNeighbors(vertex);
        for (Edge neighbor : neighbors) {
            if (!visited.contains(neighbor.vertex)) {
                dfsRecursive(neighbor.vertex, visited, result);
            }
        }
    }

    public List<String> bfs(String startVertex) {
        List<String> result = new ArrayList<>();
        Set<String> visited = new HashSet<>();
        java.util.Queue<String> queue = new java.util.LinkedList<>();
        
        queue.offer(startVertex);
        visited.add(startVertex);

        while (!queue.isEmpty()) {
            String vertex = queue.poll();
            result.add(vertex);

            List<Edge> neighbors = getNeighbors(vertex);
            for (Edge neighbor : neighbors) {
                if (!visited.contains(neighbor.vertex)) {
                    visited.add(neighbor.vertex);
                    queue.offer(neighbor.vertex);
                }
            }
        }

        return result;
    }

    public boolean hasPath(String startVertex, String endVertex) {
        Set<String> visited = new HashSet<>();
        java.util.Queue<String> queue = new java.util.LinkedList<>();
        
        queue.offer(startVertex);
        visited.add(startVertex);

        while (!queue.isEmpty()) {
            String vertex = queue.poll();

            if (vertex.equals(endVertex)) {
                return true;
            }

            List<Edge> neighbors = getNeighbors(vertex);
            for (Edge neighbor : neighbors) {
                if (!visited.contains(neighbor.vertex)) {
                    visited.add(neighbor.vertex);
                    queue.offer(neighbor.vertex);
                }
            }
        }

        return false;
    }

    public List<String> getVertices() {
        return new ArrayList<>(vertices.keySet());
    }

    public int getVertexCount() {
        return vertices.size();
    }

    public boolean isEmpty() {
        return vertices.isEmpty();
    }

    public void clear() {
        vertices.clear();
    }

    // Usage Example
    public static void main(String[] args) {
        Graph graph = new Graph(false); // Undirected graph

        graph.addVertex("A");
        graph.addVertex("B");
        graph.addVertex("C");
        graph.addVertex("D");
        graph.addEdge("A", "B");
        graph.addEdge("A", "C");
        graph.addEdge("B", "D");
        graph.addEdge("C", "D");

        System.out.println("Vertices: " + graph.getVertices());
        System.out.println("DFS: " + graph.dfs("A"));
        System.out.println("BFS: " + graph.bfs("A"));
        System.out.println("Has path A to D: " + graph.hasPath("A", "D"));
    }
}

